### PR TITLE
[BugFix] Adjust NULL_TYPE hack processing(2)

### PR DESCRIPTION
**Pull request by liuyehcf**
Signed-off-by: liuyehcf <1559500551@qq.com>

## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #https://github.com/StarRocks/StarRocksTest/issues/1853
Fixes #https://github.com/StarRocks/StarRocksTest/issues/1767

## Hack TYPE_NULL for all branches in ScalarOperatorToExpr

This pr is a contiunation of preous work #17107, in which we only handle `visitArray` `visitVariableReference` and `visitConstant`. But other branches also need the hacking process as well, so I make it up in this pr.

## Make sure MapColumn' key/value, ArrayColumn's element is nullable

For ArrayColumn, element must be nullable to facilitate handling the nested types. and key/value for MapColumn as well.

But, some place will pass non-nullable column to ArrayColumn/MapColumn's ctor, so I add an DCHECK and modify the realted tests.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
starrocks_fe_unittest

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 25 / 27 (92.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|20|22|90.91%|[164, 199]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 46 / 50 (92.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|35|39|89.74%|[203, 206, 207, 208]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 37 / 41 (90.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|32|36|88.89%|[200, 203, 204, 205]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by liuyehcf** [comment]
run all

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 37 / 41 (90.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|32|36|88.89%|[200, 203, 204, 205]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run all

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 37 / 41 (90.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|32|36|88.89%|[200, 203, 204, 205]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_be_unittest

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_be_unittest

**Comment by liuyehcf** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 40 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|31|35|88.57%|[200, 203, 204, 205]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Review by packy92 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 40 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|31|35|88.57%|[200, 203, 204, 205]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/analysis/CloneExpr.java [hunk 1]
```diff
@@ -37,6 +37,12 @@ public Type getType() {
         return getChild(0).getType();
     }
 
+    @Override
+    public void setType(Type type) {
+        super.setType(type);
+        getChild(0).setType(type);
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/analysis/CloneExpr.java:43]: cloneExpr has exactly one child.

**Review by kangkaisen — DISMISSED** [review]

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 40 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|31|35|88.57%|[200, 203, 204, 205]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 40 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|31|35|88.57%|[200, 203, 204, 205]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/CloneExpr.java|3|3|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_be_unittest

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Pull request merged**